backend/src/stwo/prover.rs

```diff
@@ -1,95 +1,315 @@
+                                .iter()
+                                .filter(|&size| size.is_power_of_two())
+                                .map(|size| {
+                                    let twiddles = B::precompute_twiddles(
+                                        CanonicCoset::new(size.ilog2() + 1 + FRI_LOG_BLOWUP as u32)
+                                            .circle_domain()
+                                            .half_coset,
+                                    );
+                                    (size as usize, twiddles)
+                                })
+                                .collect::<Vec<_>>()
+                        })
+                    })
+                    .collect::<Vec<_>>()
+            })
+            .collect();
+        // only the first one is used, machines with varying sizes are not supported yet, and it is checked in backendfactory create function.
+        let prover_channel = &mut <MC as MerkleChannel>::C::default();
+        let mut commitment_scheme =
+            CommitmentSchemeProver::<'_, B, MC>::new(config, twiddles_map.iter().next().unwrap().1);
+
+        let mut tree_builder = commitment_scheme.tree_builder();
+
+        // Get the list of constant polynomials with next reference constraint
+        let constant_list: Vec<usize> = get_constant_with_next_list(&self.analyzed);
+
+        //commit to the constant polynomials that are without next reference constraint
+        if let Some((_, table_proving_key)) =
+            self.proving_key
+                .preprocessed
+                .as_ref()
+                .and_then(|preprocessed| {
+                    preprocessed
+                        .iter()
+                        .find_map(|(_, table_collection)| table_collection.iter().next())
+                })
+        {
+            tree_builder.extend_evals(
+                table_proving_key
+                    .constant_trace_circle_domain
+                    .clone()
+                    .into_iter() // Convert it into an iterator
+                    .enumerate() // Enumerate to get (index, value)
+                    .filter(|(index, _)| !constant_list.contains(index)) // Keep only elements whose index is not in `constant_list`
+                    .map(|(_, element)| element),
+            );
+        } else {
+            tree_builder.extend_evals([]);
+        }
+        tree_builder.commit(prover_channel);
+
+        let mut trace: ColumnVec<CircleEvaluation<B, BaseField, BitReversedOrder>> = witness
+            .iter()
+            .map(|(_name, values)| {
+                gen_stwo_circle_column::<F, B, M31>(
+                    *domain_map.get(&(values.len().ilog2() as usize)).unwrap(),
+                    values,
+                )
+            })
+            .collect();
 
-        let trace = gen_stwo_circuit_trace::<F, B, M31>(witness);
+        //extend the witness trace with the constant polys that have next reference constraint
```

Comment:
I have a doubt about this: what prevents the prover from setting these to arbitrary values? What I was rather expecting, would be to commit at setup phase to the shifted fixed columns, with an optimization being to only commit to what is being accessed in the constraints.

So basically having the proving key contain the non shifted and shifted fixed columns, instead of only the non-shifted fixed columns now.